In [159]:
import math
import csv
import pandas as pd
import numpy as np

In [160]:
df1=pd.read_csv("AAPL_weekly_return_volatility.csv")
df1

Year  Week_Number  mean_return  volatility
0    2017            1      0.37800    0.560748
1    2017            2      0.19200    0.538286
2    2017            3      0.20175    0.428707
3    2017            4      0.32480    0.711057
4    2017            5      1.17260    2.767879
..    ...          ...          ...         ...
258  2021           48      0.64780    1.904606
259  2021           49      2.09540    1.445607
260  2021           50     -0.91880    2.482887
261  2021           51      0.74825    1.230668
262  2021           52      0.59000    1.511181

[263 rows x 4 columns]

In [161]:
df1['label']=""
for i in df1.index:
    if df1['mean_return'][i]<0.1 and df1['volatility'][i]>1:
        df1['label'][i+1]="red"
    else:
        df1['label'][i+1]="green"
df1
df1['Label']=0
for i in df1.index:
    if df1['label'][i]=="red":
        df1['Label'][i]=0
    else:
        df1['Label'][i]=1
df1

<ipython-input-161-98e4bc7a3a49>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['label'][i+1]="green"
<ipython-input-161-98e4bc7a3a49>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['label'][i+1]="red"
<ipython-input-161-98e4bc7a3a49>:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df1['Label'][i]=1
<ipython-input-161-98e4bc7a3a49>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFr

Year  Week_Number  mean_return  volatility  label  Label
0    2017            1      0.37800    0.560748             1
1    2017            2      0.19200    0.538286  green      1
2    2017            3      0.20175    0.428707  green      1
3    2017            4      0.32480    0.711057  green      1
4    2017            5      1.17260    2.767879  green      1
..    ...          ...          ...         ...    ...    ...
258  2021           48      0.64780    1.904606    red      0
259  2021           49      2.09540    1.445607  green      1
260  2021           50     -0.91880    2.482887  green      1
261  2021           51      0.74825    1.230668    red      0
262  2021           52      0.59000    1.511181  green      1

[263 rows x 6 columns]

In [162]:
df=pd.read_csv("AAPL_weekly_return_detailed.csv")

In [163]:
df['label'] = ""
maps = {}
for ind in df1.index:
    year = df1['Year'][ind]
    week_no = df1['Week_Number'][ind]
    if year not in maps:
        maps[year] = {}
    if week_no not in maps[year]:
        maps[year][week_no] = df1['label'][ind]

#df['label'] = ""
for ind in df.index:
    df['label'][ind] = maps[df['Year'][ind]][df['Week_Number'][ind]]

<ipython-input-163-49b28965544d>:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['label'][ind] = maps[df['Year'][ind]][df['Week_Number'][ind]]


In [164]:
df2=df1[df1['Year']==2020]
df2

Year  Week_Number  mean_return  volatility  label  Label
158  2020            0      0.65500    2.300925  green      1
159  2020            1      0.85720    1.040851  green      1
160  2020            2      0.54340    1.404286  green      1
161  2020            3     -0.03175    0.547226  green      1
162  2020            4     -0.51960    3.137070  green      1
163  2020            5      0.73040    1.745337    red      0
164  2020            6      0.31200    1.250071  green      1
165  2020            7     -0.91800    1.658535  green      1
166  2020            8     -2.62920    3.343723    red      0
167  2020            9      1.24000    5.542410    red      0
168  2020           10     -0.41480    9.574478  green      1
169  2020           11     -3.60620    6.455767    red      0
170  2020           12      1.69600    5.829284    red      0
171  2020           13     -0.47600    3.146157  green      1
172  2020           14      2.71175    4.285812    red      0
173  2020           15      1.10760    2.575317  green      1
174  2020           16      0.04240    2.767435  green      1
175  2020           17      0.44700    2.206188    red      0
176  2020           18      1.47240    0.550908  green      1
177  2020           19     -0.15060    1.209716  green      1
178  2020           20      0.72420    1.415695    red      0
179  2020           21     -0.07350    0.461216  green      1
180  2020           22      0.84600    1.351260  green      1
181  2020           23      0.47680    3.146437  green      1
182  2020           24      0.64340    1.306696  green      1
183  2020           25      0.24800    2.520517  green      1
184  2020           26      0.73775    1.135616  green      1
185  2020           27      1.05980    1.348667  green      1
186  2020           28      0.08980    1.111388  green      1
187  2020           29     -0.75840    2.466230    red      0
188  2020           30      2.86460    4.528400    red      0
189  2020           31      0.95300    2.220969  green      1
190  2020           32      0.69660    2.382896  green      1
191  2020           33      1.61380    2.192787  green      1
192  2020           34      0.07580    1.159521  green      1
193  2020           35     -0.52760    4.857081    red      0
194  2020           36     -1.82950    4.479117    red      0
195  2020           37     -0.91260    2.558190    red      0
196  2020           38      1.03780    3.123115    red      0
197  2020           39      0.15120    2.211689  green      1
198  2020           40      0.71060    2.296290  green      1
199  2020           41      0.39540    3.490233  green      1
200  2020           42     -0.67020    1.379075  green      1
201  2020           43     -1.03440    3.969262    red      0
202  2020           44      1.79420    1.971979    red      0
203  2020           45      0.10900    1.822563  green      1
204  2020           46     -0.32040    0.946634  green      1
205  2020           47     -0.14625    1.905579  green      1
206  2020           48      0.96220    1.560562    red      0
207  2020           49      0.03440    1.415744  green      1
208  2020           50      0.71000    2.541381    red      0
209  2020           51      1.03975    1.459982  green      1
210  2020           52      0.15550    2.294409  green      1

In [165]:
#X=df2.drop(['class','Date','Weekday','label','Year','Day'],axis=1)
#X


In [166]:
X=df2.drop(['Year','Week_Number','label','Label'],axis=1)
#X

In [167]:
Y=df2['Label']

In [168]:
from sklearn.model_selection import train_test_split  
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.50) 

In [169]:

from sklearn.linear_model import LogisticRegression

classifier = LogisticRegression(random_state=0)
classifier.fit(X_train, Y_train)

LogisticRegression(random_state=0)

In [213]:
y_pred = classifier.predict(X_test)
acc3 =  classifier.score(X_test, Y_test)
print(acc3)

0.8461538461538461


array([1.67588103])

In [214]:
print(classifier.coef_)
print(classifier.intercept_)

[[ 0.01762462 -0.36021246]]
[1.67588103]


#question1

Equation of logistic regression for year one is:

y= 0.0176x1 -0.3602x2 + 1.67588

In [173]:
df3=df1[df1['Year']==2021]
#df3

In [181]:
#predict labels
df['class']=0
for i in df.index:
    if df['label'][i]=="green":
        df['class'][i]=1
    else:
        df['class'][i]=0
X = df[['mean_return' , 'volatility']].values
Y = df[['class']].values
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.5)
log_classifier = LogisticRegression(random_state = 0)
log_classifier.fit(X_train , Y_train)
pred_k = log_classifier.predict(X_test)
y_pred = classifier.predict(X_test)
acc =  classifier.score(X_test, Y_test)


<ipython-input-181-611155d607dc>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['class'][i]=0
<ipython-input-181-611155d607dc>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['class'][i]=1
C:\Users\harib\anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


Open        High         Low       Close   Adj Close     Volume  \
1007  133.520004  133.610001  126.760002  129.410004  128.087067  143301900   
1008  128.889999  131.740005  128.429993  131.009995  129.670715   97664900   
1009  127.720001  131.050003  126.379997  126.599998  125.305794  155088000   
1010  128.360001  131.630005  127.860001  130.919998  129.581635  109578200   
1011  132.429993  132.630005  130.229996  132.050003  130.700089  105158200   
...          ...         ...         ...         ...         ...        ...   
1252  173.039993  175.860001  172.149994  175.639999  174.916183   92135300   
1253  175.850006  176.850006  175.270004  176.279999  175.553558   68356600   
1254  177.089996  180.419998  177.070007  180.330002  179.586868   74919600   
1255  180.160004  181.330002  178.529999  179.289993  178.551132   79144300   
1256  179.330002  180.630005  178.139999  179.380005  178.640778   62348900   

      Return        Date  Week_Number  Year  Day    Weekday  mean_return  \
1007  -2.472  2021-01-04            1  2021    4     Monday     -0.06540   
1008   1.236  2021-01-05            1  2021    5    Tuesday     -0.06540   
1009  -3.366  2021-01-06            1  2021    6  Wednesday     -0.06540   
1010   3.412  2021-01-07            1  2021    7   Thursday     -0.06540   
1011   0.863  2021-01-08            1  2021    8     Friday     -0.06540   
...      ...         ...          ...   ...  ...        ...          ...   
1252   1.532  2021-12-22           51  2021   22  Wednesday      0.74825   
1253   0.364  2021-12-23           51  2021   23   Thursday      0.74825   
1254   2.297  2021-12-27           52  2021   27     Monday      0.59000   
1255  -0.577  2021-12-28           52  2021   28    Tuesday      0.59000   
1256   0.050  2021-12-29           52  2021   29  Wednesday      0.59000   

      volatility  label  class  
1007    2.798820  green      1  
1008    2.798820  green      1  
1009    2.798820  green      1  
1010    2.798820  green      1  
1011    2.798820  green      1  
...          ...    ...    ...  
1252    1.230668    red      0  
1253    1.230668    red      0  
1254    1.511181  green      1  
1255    1.511181  green      1  
1256    1.511181  green      1  

[250 rows x 16 columns]

In [185]:
#predicted labels
df4 = df3.loc[(df3['Year'] == 2021)]
df4['predicted']=0
a = []
for i in range(0,len(df4)):
    a.append([df4.iloc[i,2],df4.iloc[i,3]])
pred = log_classifier.predict(a)
a=list(pred)


In [186]:
for i in range(len(df4)):
    #print(i)
    df4.iloc[i,6]=pred[i]
    

In [187]:
df4['Predicted']=""
for i in df4.index:
    if df4['predicted'][i]==1:
        df4['Predicted'][i]="green"
    else:
        df4['Predicted'][i]="red"

<ipython-input-187-fa3f13adbd9e>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df4['Predicted'][i]="red"
<ipython-input-187-fa3f13adbd9e>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df4['Predicted'][i]="green"


In [188]:
flag=0
for i in df4.index:
    if df4['Label'][i]==df4['predicted'][i]:
        flag+=1
flag

34

In [191]:
print("accuracy of the predicted labels =", 34/52)

accuracy of the predicted labels = 0.6538461538461539


In [196]:
df4

Year  Week_Number  mean_return  volatility  label  Label  predicted  \
211  2021            1     -0.06540    2.798820  green      1          0   
212  2021            2     -0.74560    1.537372    red      0          1   
213  2021            3      2.27550    1.460941    red      0          0   
214  2021            4     -1.01460    2.710091  green      1          1   
215  2021            5      0.75480    1.380182    red      0          1   
216  2021            6     -0.20340    0.358251  green      1          1   
217  2021            7     -1.02875    0.862577  green      1          1   
218  2021            8     -1.35020    1.738373  green      1          1   
219  2021            9      0.06860    3.277750    red      0          0   
220  2021           10     -0.02640    3.083431    red      0          0   
221  2021           11     -0.15280    2.213869    red      0          1   
222  2021           12      0.21520    1.782412    red      0          1   
223  2021           13      0.37375    1.287959  green      1          1   
224  2021           14      1.57820    0.830448  green      1          1   
225  2021           15      0.18820    1.886390  green      1          1   
226  2021           16      0.03040    1.285043  green      1          1   
227  2021           17     -0.42760    0.687936    red      0          1   
228  2021           18     -0.14160    1.940416  green      1          1   
229  2021           19     -0.40780    2.221572    red      0          1   
230  2021           20     -0.31100    1.436281    red      0          1   
231  2021           21     -0.12760    0.940105    red      0          1   
232  2021           22      0.26250    1.327003  green      1          1   
233  2021           23      0.23280    0.685261  green      1          1   
234  2021           24      0.49160    1.414945  green      1          1   
235  2021           25      0.40580    0.854258  green      1          1   
236  2021           26      1.01060    0.688673  green      1          1   
237  2021           27      0.91325    1.239048  green      1          1   
238  2021           28      0.18440    1.468083  green      1          1   
239  2021           29      0.31100    2.008135  green      1          1   
240  2021           30     -0.36300    0.916924  green      1          1   
241  2021           31      0.07020    0.696021  green      1          1   
242  2021           32      0.40560    0.956122  green      1          1   
243  2021           33     -0.11240    1.560422  green      1          1   
244  2021           34      0.05780    0.801541    red      0          1   
245  2021           35      0.76360    1.412882  green      1          1   
246  2021           36     -0.85975    1.988616  green      1          1   
247  2021           37     -0.38980    1.009416    red      0          1   
248  2021           38      0.12540    1.405595    red      0          1   
249  2021           39     -0.58100    1.326391  green      1          1   
250  2021           40      0.04440    1.528752    red      0          1   
251  2021           41      0.27540    1.149908    red      0          1   
252  2021           42      0.52880    0.819141  green      1          1   
253  2021           43      0.15820    1.559403  green      1          1   
254  2021           44      0.22780    0.665820  green      1          1   
255  2021           45     -0.16500    1.220756  green      1          1   
256  2021           46      1.37500    1.088524    red      0          1   
257  2021           47     -0.57625    1.728194  green      1          1   
258  2021           48      0.64780    1.904606    red      0          1   
259  2021           49      2.09540    1.445607  green      1          0   
260  2021           50     -0.91880    2.482887  green      1          1   
261  2021           51      0.74825    1.230668    red      0          1   
262  2021           52      0.59000    1.511181  gree

In [206]:
X = df4[['mean_return' , 'volatility']].values
Y = df4[['predicted']].values
#for year 2
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.5)
log_classifier = LogisticRegression(random_state = 0)
log_classifier.fit(X_train , Y_train)
#pred_k = log_classifier.predict(X_test)
y_pred = log_classifier.predict(X_test)
acc =  classifier.score(X_test, Y_test)
print("accuracy of year 2",acc)

accuracy of year 2 0.8461538461538461


C:\Users\harib\anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


In [198]:
#question 3
cm = metrics.confusion_matrix(df4['Label'], df4['predicted'])
cm

array([[ 3, 16],
       [ 2, 31]], dtype=int64)

In [1]:
tpr=3/16
tnr=31/33
print(tpr,tnr)

0.1875 0.9393939393939394


In [199]:
#question 4
from sklearn import metrics
from sklearn.metrics import classification_report,confusion_matrix
classification_report(df4['Label'],df4['predicted'])

'              precision    recall  f1-score   support\n\n           0       0.60      0.16      0.25        19\n           1       0.66      0.94      0.78        33\n\n    accuracy                           0.65        52\n   macro avg       0.63      0.55      0.51        52\nweighted avg       0.64      0.65      0.58        52\n'

In [200]:
df5=df[df['Year']==2021]

In [201]:
df5

Open        High         Low       Close   Adj Close     Volume  \
1007  133.520004  133.610001  126.760002  129.410004  128.087067  143301900   
1008  128.889999  131.740005  128.429993  131.009995  129.670715   97664900   
1009  127.720001  131.050003  126.379997  126.599998  125.305794  155088000   
1010  128.360001  131.630005  127.860001  130.919998  129.581635  109578200   
1011  132.429993  132.630005  130.229996  132.050003  130.700089  105158200   
...          ...         ...         ...         ...         ...        ...   
1252  173.039993  175.860001  172.149994  175.639999  174.916183   92135300   
1253  175.850006  176.850006  175.270004  176.279999  175.553558   68356600   
1254  177.089996  180.419998  177.070007  180.330002  179.586868   74919600   
1255  180.160004  181.330002  178.529999  179.289993  178.551132   79144300   
1256  179.330002  180.630005  178.139999  179.380005  178.640778   62348900   

      Return        Date  Week_Number  Year  Day    Weekday  mean_return  \
1007  -2.472  2021-01-04            1  2021    4     Monday     -0.06540   
1008   1.236  2021-01-05            1  2021    5    Tuesday     -0.06540   
1009  -3.366  2021-01-06            1  2021    6  Wednesday     -0.06540   
1010   3.412  2021-01-07            1  2021    7   Thursday     -0.06540   
1011   0.863  2021-01-08            1  2021    8     Friday     -0.06540   
...      ...         ...          ...   ...  ...        ...          ...   
1252   1.532  2021-12-22           51  2021   22  Wednesday      0.74825   
1253   0.364  2021-12-23           51  2021   23   Thursday      0.74825   
1254   2.297  2021-12-27           52  2021   27     Monday      0.59000   
1255  -0.577  2021-12-28           52  2021   28    Tuesday      0.59000   
1256   0.050  2021-12-29           52  2021   29  Wednesday      0.59000   

      volatility  label  class  
1007    2.798820  green      1  
1008    2.798820  green      1  
1009    2.798820  green      1  
1010    2.798820  green      1  
1011    2.798820  green      1  
...          ...    ...    ...  
1252    1.230668    red      0  
1253    1.230668    red      0  
1254    1.511181  green      1  
1255    1.511181  green      1  
1256    1.511181  green      1  

[250 rows x 16 columns]

In [202]:
#question 5
df5=df[df['Year']==2021]
df5["Predicted"]=""
maps = {}
for ind in df4.index:
    year = df4['Year'][ind]
    week_no = df4['Week_Number'][ind]
    if year not in maps:
        maps[year] = {}
    if week_no not in maps[year]:
        maps[year][week_no] = df4['Predicted'][ind]

#df['label'] = ""
for ind in df5.index:
    df5['Predicted'][ind] = maps[df5['Year'][ind]][df5['Week_Number'][ind]]

<ipython-input-202-7ca69ea4c4c5>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df5["Predicted"]=""
<ipython-input-202-7ca69ea4c4c5>:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df5['Predicted'][ind] = maps[df5['Year'][ind]][df5['Week_Number'][ind]]
C:\Users\harib\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3418: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

In [203]:
df5

Open        High         Low       Close   Adj Close     Volume  \
1007  133.520004  133.610001  126.760002  129.410004  128.087067  143301900   
1008  128.889999  131.740005  128.429993  131.009995  129.670715   97664900   
1009  127.720001  131.050003  126.379997  126.599998  125.305794  155088000   
1010  128.360001  131.630005  127.860001  130.919998  129.581635  109578200   
1011  132.429993  132.630005  130.229996  132.050003  130.700089  105158200   
...          ...         ...         ...         ...         ...        ...   
1252  173.039993  175.860001  172.149994  175.639999  174.916183   92135300   
1253  175.850006  176.850006  175.270004  176.279999  175.553558   68356600   
1254  177.089996  180.419998  177.070007  180.330002  179.586868   74919600   
1255  180.160004  181.330002  178.529999  179.289993  178.551132   79144300   
1256  179.330002  180.630005  178.139999  179.380005  178.640778   62348900   

      Return        Date  Week_Number  Year  Day    Weekday  mean_return  \
1007  -2.472  2021-01-04            1  2021    4     Monday     -0.06540   
1008   1.236  2021-01-05            1  2021    5    Tuesday     -0.06540   
1009  -3.366  2021-01-06            1  2021    6  Wednesday     -0.06540   
1010   3.412  2021-01-07            1  2021    7   Thursday     -0.06540   
1011   0.863  2021-01-08            1  2021    8     Friday     -0.06540   
...      ...         ...          ...   ...  ...        ...          ...   
1252   1.532  2021-12-22           51  2021   22  Wednesday      0.74825   
1253   0.364  2021-12-23           51  2021   23   Thursday      0.74825   
1254   2.297  2021-12-27           52  2021   27     Monday      0.59000   
1255  -0.577  2021-12-28           52  2021   28    Tuesday      0.59000   
1256   0.050  2021-12-29           52  2021   29  Wednesday      0.59000   

      volatility  label  class Predicted  
1007    2.798820  green      1       red  
1008    2.798820  green      1       red  
1009    2.798820  green      1       red  
1010    2.798820  green      1       red  
1011    2.798820  green      1       red  
...          ...    ...    ...       ...  
1252    1.230668    red      0     green  
1253    1.230668    red      0     green  
1254    1.511181  green      1     green  
1255    1.511181  green      1     green  
1256    1.511181  green      1     green  

[250 rows x 17 columns]

In [204]:
#trading strategy (buy and hold)
#buy and hold
m=100
bh_2018 = df5.loc[df5['Year'] == 2021]
bh_2018['invest']= np.where(bh_2018['Year'] == 2021,0.0,0.0)
for ind in bh_2018.index:
    if ind == bh_2018.first_valid_index():
        s = round((m/bh_2018['Open'][ind]),2)
    bh_2018['invest'][ind] = s*bh_2018['Close'][ind]
bh_2018


<ipython-input-204-2dca04cb8d69>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  bh_2018['invest'][ind] = s*bh_2018['Close'][ind]


Open        High         Low       Close   Adj Close     Volume  \
1007  133.520004  133.610001  126.760002  129.410004  128.087067  143301900   
1008  128.889999  131.740005  128.429993  131.009995  129.670715   97664900   
1009  127.720001  131.050003  126.379997  126.599998  125.305794  155088000   
1010  128.360001  131.630005  127.860001  130.919998  129.581635  109578200   
1011  132.429993  132.630005  130.229996  132.050003  130.700089  105158200   
...          ...         ...         ...         ...         ...        ...   
1252  173.039993  175.860001  172.149994  175.639999  174.916183   92135300   
1253  175.850006  176.850006  175.270004  176.279999  175.553558   68356600   
1254  177.089996  180.419998  177.070007  180.330002  179.586868   74919600   
1255  180.160004  181.330002  178.529999  179.289993  178.551132   79144300   
1256  179.330002  180.630005  178.139999  179.380005  178.640778   62348900   

      Return        Date  Week_Number  Year  Day    Weekday  mean_return  \
1007  -2.472  2021-01-04            1  2021    4     Monday     -0.06540   
1008   1.236  2021-01-05            1  2021    5    Tuesday     -0.06540   
1009  -3.366  2021-01-06            1  2021    6  Wednesday     -0.06540   
1010   3.412  2021-01-07            1  2021    7   Thursday     -0.06540   
1011   0.863  2021-01-08            1  2021    8     Friday     -0.06540   
...      ...         ...          ...   ...  ...        ...          ...   
1252   1.532  2021-12-22           51  2021   22  Wednesday      0.74825   
1253   0.364  2021-12-23           51  2021   23   Thursday      0.74825   
1254   2.297  2021-12-27           52  2021   27     Monday      0.59000   
1255  -0.577  2021-12-28           52  2021   28    Tuesday      0.59000   
1256   0.050  2021-12-29           52  2021   29  Wednesday      0.59000   

      volatility  label  class Predicted      invest  
1007    2.798820  green      1       red   97.057503  
1008    2.798820  green      1       red   98.257496  
1009    2.798820  green      1       red   94.949999  
1010    2.798820  green      1       red   98.189999  
1011    2.798820  green      1       red   99.037502  
...          ...    ...    ...       ...         ...  
1252    1.230668    red      0     green  131.730000  
1253    1.230668    red      0     green  132.209999  
1254    1.511181  green      1     green  135.247501  
1255    1.511181  green      1     green  134.467495  
1256    1.511181  green      1     green  134.535004  

[250 rows x 18 columns]

In [205]:
df=pd.read_csv("AAPL_weekly_return_detailed.csv")
cash=100
quantity=0
df['invest']=0.0
df['profit']=0.0
df['decision']=''
df['profit/loss']=0.0
df['Overnight']=0.0
long=[]
short=[]
profit=0
a=[]
for i in df.index:
    if i == 0:
        quantity=100/df['Open'][i]
        df['invest'][i]=quantity*df['Close'][i]
        cash=df['invest'][i]
        df['profit'][i]=cash-100
        long.append(df['profit/loss'][i])
        #profit=profit+df['profit/loss'][i]
        #print(cash)
    elif df['Open'][i]>df['Close'][i-1]:
        quantity=100/df['Open'][i]
        cash=quantity*df['Close'][i]
        #print(cash)
        df['invest'][i]=cash
        df['profit'][i]=cash-100
        df['decision'][i]='buy'
        df['profit/loss'][i]=((df['Close'][i-1]/df['Open'][i])*(df['Close'][i]-df['Open'][i]))
        long.append(df['profit/loss'][i])
        profit=profit+df['profit'][i]
        diff = round(((df['Open'][i] - df['Close'][i-1])/df['Close'][i-1])*100 ,2)
        df['Overnight'][i]=diff
        #print(cash)

    elif cash>100 and df['Close'][i-1]>df['Open'][i] :
        #print("3")
        #if df['Close'][i-1]>df['Open'][i]:
        buy1=100/df['Open'][i]
        sell=buy1*df['Close'][i]
        a=(buy1*df['Open'][i])-sell
        #print(a)
        df['invest'][i]=sell
        df['profit'][i]=a
        df['decision'][i]='sell short'
        df['profit/loss'][i]=((df['Close'][i-1]/df['Open'][i])*(df['Close'][i]-df['Open'][i]))
        short.append(df['profit/loss'][i])
        profit=profit+df['profit'][i]
        diff = round(((df['Open'][i] - df['Close'][i-1])/df['Close'][i-1])*100 ,2)
        df['Overnight'][i]=diff
        #print(cash)
        
    elif df['Close'][i-1]==df['Open'][i]:
        df['invest'][i]=100
        df['profit'][i]=0
        df['decision'][i]='Dont do anything'
        df['profit/loss'][i]=0
        long.append(df['profit/loss'][i])
        profit=profit+df['profit'][i]
        diff = round(((df['Open'][i] - df['Close'][i-1])/df['Close'][i-1])*100 ,2)
        df['Overnight'][i]=diff
    else:
        if df['Close'][i-1]>df['Open'][i]:
            buy1=100/df['Open'][i]
            sell=buy1*df['Close'][i]
            a=(buy1*df['Open'][i])-sell
            #print(a)
            df['invest'][i]=sell
            df['profit'][i]=a
            df['decision'][i]='sell short'
            df['profit/loss'][i]=((df['Close'][i-1]/df['Open'][i])*(df['Close'][i]-df['Open'][i]))
            short.append(df['profit/loss'][i])
            #print(cash)
            profit=profit+df['profit'][i]
            diff = round(((df['Open'][i] - df['Close'][i-1])/df['Close'][i-1])*100 ,2)
            df['Overnight'][i]=diff
            
df


<ipython-input-205-e85f39487897>:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['invest'][i]=quantity*df['Close'][i]
<ipython-input-205-e85f39487897>:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['profit'][i]=cash-100
<ipython-input-205-e85f39487897>:43: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['invest'][i]=sell
<ipython-input-205-e85f39487897>:44: SettingWithCopyWarning: 
A value is trying to be set on a copy of a

Open        High         Low       Close   Adj Close     Volume  \
0      28.950001   29.082500   28.690001   29.037500   27.219831  115127600   
1      28.962500   29.127501   28.937500   29.004999   27.189365   84472400   
2      28.980000   29.215000   28.952499   29.152500   27.327635   88774400   
3      29.195000   29.540001   29.117500   29.477501   27.632290  127007600   
4      29.487499   29.857500   29.485001   29.747499   27.885387  134247600   
...          ...         ...         ...         ...         ...        ...   
1252  173.039993  175.860001  172.149994  175.639999  174.916183   92135300   
1253  175.850006  176.850006  175.270004  176.279999  175.553558   68356600   
1254  177.089996  180.419998  177.070007  180.330002  179.586868   74919600   
1255  180.160004  181.330002  178.529999  179.289993  178.551132   79144300   
1256  179.330002  180.630005  178.139999  179.380005  178.640778   62348900   

      Return        Date  Week_Number  Year  Day    Weekday  mean_return  \
0      0.000  2017-01-03            1  2017    3    Tuesday      0.37800   
1     -0.112  2017-01-04            1  2017    4  Wednesday      0.37800   
2      0.509  2017-01-05            1  2017    5   Thursday      0.37800   
3      1.115  2017-01-06            1  2017    6     Friday      0.37800   
4      0.916  2017-01-09            2  2017    9     Monday      0.19200   
...      ...         ...          ...   ...  ...        ...          ...   
1252   1.532  2021-12-22           51  2021   22  Wednesday      0.74825   
1253   0.364  2021-12-23           51  2021   23   Thursday      0.74825   
1254   2.297  2021-12-27           52  2021   27     Monday      0.59000   
1255  -0.577  2021-12-28           52  2021   28    Tuesday      0.59000   
1256   0.050  2021-12-29           52  2021   29  Wednesday      0.59000   

      volatility      invest    profit    decision  profit/loss  Overnight  
0       0.560748  100.302244  0.302244                 0.000000       0.00  
1       0.560748  100.146740 -0.146740  sell short     0.042610      -0.26  
2       0.560748  100.595240 -0.595240  sell short     0.172649      -0.09  
3       0.560748  100.967636  0.967636         buy     0.282090       0.15  
4       0.538286  100.881730  0.881730         buy     0.259912       0.03  
...          ...         ...       ...         ...          ...        ...  
1252    1.230668  101.502546  1.502546         buy     2.599255       0.03  
1253    1.230668  100.244522  0.244522         buy     0.429479       0.12  
1254    1.511181  101.829581  1.829581         buy     3.225186       0.46  
1255    1.511181   99.517090  0.482910  sell short    -0.870831      -0.09  
1256    1.511181  100.027883  0.027883         buy     0.049992       0.02  

[1257 rows x 19 columns]